File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/AddSourceMenu.swift
internal import SwiftUI

internal import StoryArcCore

/// The five ways to bring something in to read.
///
/// A menu rather than five buttons. There are four kinds of library now and there will be
/// more; a toolbar with one button per kind would crowd out the controls a reader uses every
/// day. Android's `AddSourceMenu` is the same menu.
///
/// It was written, translated, and mounted by nothing: `LibraryToolbar` hand-built a
/// four-item menu of its own that omitted the import, so the only way a file reached
/// StoryArc on iOS was the system's own Open-in handler. Two menus for one job is how one of
/// them ends up a row short, which is exactly what happened.
struct AddSourceMenu: View {
    let addFolder: () -> Void
    /// `local-library`: a file brought in from elsewhere is copied into storage the app
    /// owns. Its own control rather than a mode of "add a folder", because a reader adding a
    /// folder is pointing at something they keep, and a reader importing is handing it over.
    let importFile: () -> Void
    let addCatalogue: () -> Void
    let addKavita: () -> Void
    let addShare: () -> Void

    var body: some View {
        Menu {
            item(.localFolder, addFolder)
            ImportPublicationButton(action: importFile)
            item(.opdsCatalog, addCatalogue)
            item(.kavitaServer, addKavita)
            item(.networkShare, addShare)
        } label: {
            Label {
                Text("library.addSource", bundle: .module)
            } icon: {
                Image(systemName: "plus")
            }
        }
    }

    /// One kind of place, named and explained.
    ///
    /// `sources` asks for "a one-line explanation of each", and the live delta puts that
    /// naming here rather than on the first screen — this menu is the secondary action, and
    /// choosing between the four is the question it asks. The sentences were written and
    /// translated into four languages and drawn by nobody: the rows used to carry the four
    /// *sheets'* titles, which name the destination and say nothing about what it is.
    ///
    /// A second `Text` in a menu button's label is how SwiftUI draws a subtitle, which is the
    /// platform's own shape for exactly this. The kind carries both keys and the symbol, from
    /// ``SourceKind`` in `SourcePresentation.swift`, so a fifth kind is named everywhere at
    /// once rather than in whichever menu was remembered.
    @ViewBuilder
    private func item(_ kind: SourceKind, _ action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(kind.titleKey, bundle: .module)
            Text(kind.explanationKey, bundle: .module)
            Image(systemName: kind.symbolName)
        }
    }
}

File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/AddSourceMenu.kt
package app.storyarc.feature.library

import androidx.compose.foundation.layout.Column
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.FileDownload
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.stringResource
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.model.SourceKind

/**
 * The ways to add a source that exist, behind one button.
 *
 * Its own file since the catalogue strip it used to share one with was removed: task 2.4 of
 * `one-library-three-destinations` took the per-source chips off the shelf, and this menu
 * has nothing to do with them — it is the library toolbar's way in to the add-a-source
 * sheets, and iOS keeps its own in `AddSourceMenu.swift`.
 */
@Composable
fun AddSourceMenu(
    onAddFolder: () -> Unit,
    onAddCatalogue: () -> Unit,
    onAddKavita: () -> Unit = {},
    onAddShare: () -> Unit = {},
    /**
     * Copies one publication into the app's own storage.
     *
     * Beside the source kinds rather than among them: `local-library` gives imported copies
     * a requirement of their own, and "On this device" is not a place a reader configures.
     */
    onImport: () -> Unit = {},
) {
    var open by remember { mutableStateOf(false) }

    IconButton(onClick = { open = true }) {
        Icon(
            imageVector = Icons.Filled.Add,
            contentDescription = stringResource(R.string.library_add_source),
            tint = MaterialTheme.colorScheme.primary,
        )
    }
    DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
        AddSourceItems(
            onChosen = { open = false },
            onAddFolder = onAddFolder,
            onImport = onImport,
            onAddCatalogue = onAddCatalogue,
            onAddKavita = onAddKavita,
            onAddShare = onAddShare,
        )
    }
}

/**
 * The five rows themselves, so the two menus that offer them cannot disagree.
 *
 * There are two ways in — this toolbar button and the empty state's plain secondary action —
 * and they held two copies of the same list. Two lists for one job is how one of them ends up
 * a row short, which is the defect iOS's `AddSourceMenu` was written to end.
 *
 * **Each of the four kinds is named and explained.** `sources` asks for "a one-line
 * explanation of each", and the live delta puts that naming here rather than on the first
 * screen: this is the secondary action, and choosing between the four is the question it asks.
 * The eight sentences were written and translated into four languages and drawn by nobody.
 * They come from [SourceKind.titleRes] and [SourceKind.explanationRes], so a fifth kind is
 * named in both menus at once rather than in whichever one was remembered.
 */
@Composable
internal fun AddSourceItems(
    onChosen: () -> Unit,
    onAddFolder: () -> Unit,
    onImport: () -> Unit,
    onAddCatalogue: () -> Unit,
    onAddKavita: () -> Unit,
    onAddShare: () -> Unit,
) {
    SourceKindItem(SourceKind.LOCAL_FOLDER, onChosen, onAddFolder)
    DropdownMenuItem(
        text = { Text(stringResource(R.string.library_import)) },
        leadingIcon = { Icon(Icons.Filled.FileDownload, contentDescription = null) },
        onClick = {
            onChosen()
            onImport()
        },
    )
    SourceKindItem(SourceKind.OPDS_CATALOG, onChosen, onAddCatalogue)
    SourceKindItem(SourceKind.KAVITA_SERVER, onChosen, onAddKavita)
    SourceKindItem(SourceKind.NETWORK_SHARE, onChosen, onAddShare)
}

/**
 * One kind of place: what it is called, and one line saying what it is.
 *
 * Material publishes no two-line menu row, so the two lines go in the row's own text slot —
 * the shape a `ListItem` would give, without dragging a list item into a menu. iOS draws the
 * same pair as a menu button's title and subtitle, which is that platform's own shape for it.
 */
@Composable
private fun SourceKindItem(kind: SourceKind, onChosen: () -> Unit, onClick: () -> Unit) {
    val palette = LocalStoryArcPalette.current
    DropdownMenuItem(
        text = {
            Column {
                Text(stringResource(kind.titleRes))
                Text(
                    text = stringResource(kind.explanationRes),
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.textSecondary,
                )
            }
        },
        leadingIcon = { Icon(kind.icon, contentDescription = null) },
        onClick = {
            onChosen()
            onClick()
        },
    )
}
